rust/headless-client/src/telemetry.rs

```diff
@@ -0,0 +1,150 @@
+use arc_swap::ArcSwapOption;
+use std::time::Duration;
+
+pub struct Dsn(&'static str);
+
+// TODO: Dynamic DSN
+// Sentry docs say this does not need to be protected:
+// > DSNs are safe to keep public because they only allow submission of new events and related event data; they do not allow read access to any information.
+// <https://docs.sentry.io/concepts/key-terms/dsn-explainer/#dsn-utilization>
+
+pub const GUI_DSN: Dsn = Dsn("https://2e17bf5ed24a78c0ac9e84a5de2bd6fc@o4507971108339712.ingest.us.sentry.io/4508008945549312");
+pub const HEADLESS_DSN: Dsn = Dsn("https://bc27dca8bb37be0142c48c4f89647c13@o4507971108339712.ingest.us.sentry.io/4508010028728320");
+pub const IPC_SERVICE_DSN: Dsn = Dsn("https://0590b89fd4479494a1e7ffa4dc705001@o4507971108339712.ingest.us.sentry.io/4508008896069632");
+
+#[derive(Default)]
+pub struct Telemetry {
+    /// The sentry guard itself
+    ///
+    /// `arc_swap` is used here because the borrowing for the GUI process is
+    /// complex. If the user has already consented to using telemetry, we want
+    /// to start Sentry as soon as possible inside `main`, and keep the guard
+    /// alive there. But sentry's `Drop` hook will never get called during
+    /// normal GUI operation because Tauri internally calls `std::process::exit`
+    /// to bail out of its Tao event loop. So we want the GUI controller to
+    /// be able to gracefully close down Sentry, even though it won't have
+    /// a mutable handle to it, and even though `main` is actually what holds
+    /// the handle. Wrapping it all in `ArcSwap` makes it simpler.
+    inner: ArcSwapOption<sentry::ClientInitGuard>,
+}
+
+impl Clone for Telemetry {
+    fn clone(&self) -> Self {
+        Self {
+            inner: ArcSwapOption::new(self.inner.load().clone()),
+        }
+    }
+}
+
+impl Telemetry {
+    /// Flushes events to sentry.io and drops the guard.
+    pub fn close(&self) {
+        self.stop_sentry()
+    }
+
+    /// Allows users to opt in or out arbitrarily at run time.
+    pub fn set_enabled(&self, dsn: Option<Dsn>) {
+        if let Some(dsn) = dsn {
+            self.start_sentry(dsn)
+        } else {
+            self.stop_sentry()
+        }
+    }
+
+    fn start_sentry(&self, dsn: Dsn) {
+        // Since it's `arc_swap` and not `Option`, there is a TOCTOU here,
+        // but in practice it should never trigger
+        if self.inner.load().is_some() {
+            return;
+        }
+        let inner = sentry::init((
+            dsn.0,
+            sentry::ClientOptions {
+                release: sentry::release_name!(),
+                traces_sample_rate: 1.0,
+                ..Default::default()
+            },
```

Comment:
Do we need to configure any of the [integrations](https://docs.rs/sentry/0.34.0/sentry/trait.Integration.html) here? By default there aren't any I think.

Reply:
Should we set `environment` to the API endpoint we are connecting to?

https://docs.sentry.io/platforms/rust/configuration/options/#environment